docs/src/index.md

```diff
@@ -4,6 +4,350 @@ DocTestSetup = :(using ChainRulesCore, ChainRules)
+
+Do _not_ wrap _variables_ in a `@thunk`; wrap the _computations_ that fill those variables in `@thunk`:
+
+```julia
+# good:
+∂A = @thunk(foo(x))
+return ∂A
+
+# bad:
+∂A = foo(x)
+return @thunk(∂A)
+```
+In the bad example `foo(x)` gets computed eagerly, and all that the thunk is doing is wrapping the already calculated result in a function that returns it.
+
+### Be careful with using `adjoint` when you mean `transpose`
+
+Remember for complex numbers `a'` (i.e. `adjoint(a)`) takes the complex conjugate.
+Instead you probably want `transpose(a)`.
+
+While there are arguments that for reverse-mode taking the adjoint is correct, it is not currently the behavior of ChainRules to do so.
+Feel free to open an issue to discuss it.
+
+### Code Style
+
+Use named local functions for the `pushforward`/`pullback`:
+
+```julia
+# good:
+function frule(::typeof(foo), x)
+    Y = foo(x)
+    function foo_pushforward(_, ẋ)
+        return bar(ẋ)
+    end
+    return Y, foo_pushforward
+end
+
+# bad:
+function frule(::typeof(foo), x)
+    return foo(x), (_, ẋ) -> bar(ẋ)
+end
+```
+
+While this is more verbose, it ensures that if an error is thrown during the `pullback`/`pushforward` the [`gensym`](https://docs.julialang.org/en/v1/base/base/#Base.gensym) name of the local function will include the name you gave it.
+This makes it a lot simpler to debug from the stacktrace.
+
+### Write tests
+
+There are fairly decent tools for writing tests based on [FiniteDifferences.jl](https://github.com/JuliaDiff/FiniteDifferences.jl).
+They are in [`tests/test_utils.jl`](https://github.com/JuliaDiff/ChainRules.jl/blob/master/test/test_util.jl).
+Take a look at existing test and you should see how to do stuff.
+
+!!! warning
+    Use finite differencing to test derivatives.
+    Don't use analytical derivations for derivatives in the tests!
+    Since the rules are analytic expressions, re-writing those same expressions in the tests
+    can not be an effective way to test, and will give misleading test coverage.
+
+### CAS systems are your friends.
+
+It is very easy to check gradients or derivatives with a computer algebra system (CAS) like [WolframAlpha](https://www.wolframalpha.com/input/?i=gradient+atan2%28x%2Cy%29).
+
+------------------------------------------
+
+## FAQ
+
+### What is up with the different symbols?
+
+ - `Δx` is the input to a propagator, (i.e a _seed_ for a _pullback_; or a _perturbation_ for a _pushforward_)
+ - `∂x` is the output of a propagator
+ - `dx` could be anything, including a pullback. It really should not show up outside of tests.
+ - `v̇` is a derivative of the input moving forward: ``v̇ = \frac{∂v}{∂x}`` for input ``x``, intermediate value ``v``.
+ - `v̄` is a derivative of the output moving backward: ``v̄ = \frac{∂y}{∂v}`` for output ``y``, intermediate value ``v``.
+ - `Ω` is often used as the return value of the function having the rule found for. Especially, (but not exclusively.) for scalar functions.
```

Comment:
The first sentence doesn't sound correct. And there is a dot after "exclusively".

Reply:
Better but not perfect.